--- debian/tmp/usr/lib/python2.3/site-packages/freevo/plugins/rom_drives.py	2004-01-08 10:42:40.000000000 -0500
+++ debian/tmp/usr/lib/python2.3/site-packages/freevo/plugins/rom_drives.py.new	2004-01-08 10:48:40.000000000 -0500
@@ -66,16 +66,6 @@
     # (for some strange reason, this is missing sometimes)
     CDROM_DRIVE_STATUS
 except:
-    if os.uname()[0] == 'FreeBSD':
-        # FreeBSD ioctls - there is no CDROM.py...
-        CDIOCEJECT = 0x20006318
-        CDIOCCLOSE = 0x2000631c
-        CDIOREADTOCENTRYS = 0xc0086305
-        CD_LBA_FORMAT = 1
-        CD_MSF_FORMAT = 2
-        CDS_NO_DISC = 1
-        CDS_DISC_OK = 4
-    else:
         # strange ioctrls missing
         CDROMEJECT = 0x5309
         CDROMCLOSETRAY = 0x5319
