--- freevo-1.4/src/audio/xmms.py	2003-11-17 01:41:04.000000000 +0800
+++ freevo/src/audio/xmms.py	2003-11-22 16:54:22.000000000 +0800
@@ -2,13 +2,16 @@
 # -----------------------------------------------------------------------
 # xmmsaudioplayer.py - Play music using XMMS
 # -----------------------------------------------------------------------
-# $Id$
+# $Id$
 #
 # Notes: Not working right now
 # Todo:  Integrate xmms to the new clode layout
 #
 # -----------------------------------------------------------------------
 # $Log$
 # Revision 1.1  2003/11/22 10:04:37  tcwan
 # Updated files for 1.4 final release
 #
+# Revision 1.6  2003/11/20 21:46:05  outlyer
+# Missing bracket, prevents this file from being loaded.
+#
 # Revision 1.5  2003/11/16 17:41:04  dischi
 # i18n patch from David Sagnol
 #
@@ -324,7 +327,7 @@
 	    self.mode = mode # setting global var to mode.
 	    self.repeat = repeat # Repeat playlist setting
 	    
-	    globals.skin.PopupBox(_'Lanching xmms.'))
+	    globals.skin.PopupBox(_('Lanching xmms.'))
 	    self.set_mixer_levels()
         
 	    self.thread.mode  = constants.playmode
