--- Imaging-1.1.4/Setup.in.orig	2003-09-17 10:37:24.000000000 +0800
+++ Imaging-1.1.4/Setup.in	2003-09-17 10:38:15.000000000 +0800
@@ -40,10 +40,10 @@
 # you're using a custom _tkinter instead (like in earlier versions of
 # PIL).  Edit the library specifications to match your configuration.
 
-_imagingtk  _imagingtk.c \
-	-IlibImaging Tk/tkImaging.c \
-	-I/usr/local/include -L/usr/local/lib -ltcl8.3 -ltk8.3 \
-        -L/usr/X11R6/lib -lX11 
+# _imagingtk  _imagingtk.c \
+# 	-IlibImaging Tk/tkImaging.c \
+# 	-I/usr/local/include -L/usr/local/lib -ltcl8.3 -ltk8.3 \
+#       -L/usr/X11R6/lib -lX11 
 
 # --------------------------------------------------------------------
 # Sane driver.  uncomment to enable this driver (see Sane/README for
