diff -ur SDL-1.2.5/src/video/fbcon/SDL_fbevents.c SDL-1.2.5-freevo/src/video/fbcon/SDL_fbevents.c
--- SDL-1.2.5/src/video/fbcon/SDL_fbevents.c	Wed Mar  6 05:23:03 2002
+++ SDL-1.2.5-freevo/src/video/fbcon/SDL_fbevents.c	Fri Jan 10 00:28:55 2003
@@ -153,13 +153,69 @@
 
 int FB_InGraphicsMode(_THIS)
 {
+	if (getenv("SDL_NOKEYBOARD") != NULL) {
+		return (1); /* Special handling for no keyboard */
+	}
+  
 	return((keyboard_fd >= 0) && (saved_kbd_mode >= 0));
 }
 
+
+static void
+tty_enable (void)
+{
+	int tty;
+  
+	tty = open ("/dev/tty0", O_RDWR);
+	if(tty < 0) {
+		perror("Error can't open /dev/tty0");
+		exit (1);
+	}
+
+	if (ioctl (tty, KDSETMODE, KD_TEXT) == -1) {
+		perror ("Error setting text mode for tty");
+		close (tty);
+		exit (1);
+	}
+  
+	close(tty);
+}
+
+
+static void
+tty_disable (void)
+{
+	int tty;
+
+
+	tty = open ("/dev/tty0", O_RDWR);
+	if (tty < 0) {
+		perror ("Error can't open /dev/tty0");
+		exit (1);
+	}
+
+	if (ioctl (tty, KDSETMODE, KD_GRAPHICS) == -1) {
+		perror ("Error setting graphics mode for tty");
+		close (tty);
+		exit (1);
+	}
+  
+	close (tty);
+
+}
+
+
 int FB_EnterGraphicsMode(_THIS)
 {
 	struct termios keyboard_termios;
+	int fd;
 
+        
+	if (getenv("SDL_NOKEYBOARD") != NULL) {
+		tty_disable ();
+		return (1);              /* Do not touch the keyboard */
+        }
+        
 	/* Set medium-raw keyboard mode */
 	if ( (keyboard_fd >= 0) && !FB_InGraphicsMode(this) ) {
 
@@ -219,6 +275,12 @@
 
 void FB_LeaveGraphicsMode(_THIS)
 {
+  
+	if (getenv("SDL_NOKEYBOARD") != NULL) {
+		tty_enable ();
+		return; /* Special handling for no keyboard */
+        }
+
 	if ( FB_InGraphicsMode(this) ) {
 		ioctl(keyboard_fd, KDSETMODE, KD_TEXT);
 		ioctl(keyboard_fd, KDSKBMODE, saved_kbd_mode);
@@ -841,6 +903,11 @@
 	Uint32 screen_arealen;
 	Uint8 *screen_contents;
 
+	
+	if (getenv("SDL_NOKEYBOARD") != NULL) {
+		return;        /* Do not touch the keyboard */
+	}
+        
 	/* Figure out whether or not we're switching to a new console */
 	if ( (ioctl(keyboard_fd, VT_GETSTATE, &vtstate) < 0) ||
 	     (which == vtstate.v_active) ) {
@@ -891,6 +958,11 @@
 	int scancode;
 	SDL_keysym keysym;
 
+
+	if (getenv("SDL_NOKEYBOARD") != NULL) {
+		return;        /* Do not touch the keyboard */
+	}
+        
 	nread = read(keyboard_fd, keybuf, BUFSIZ);
 	for ( i=0; i<nread; ++i ) {
 		scancode = keybuf[i] & 0x7F;
@@ -934,6 +1006,10 @@
 	int max_fd;
 	static struct timeval zero;
 
+        if (getenv("SDL_NOKEYBOARD") != NULL) {
+          return;        /* XXX Do not touch the keyboard */
+        }
+        
 	do {
 		posted = 0;
 
diff -ur SDL-1.2.5/src/video/fbcon/SDL_fbvideo.c SDL-1.2.5-freevo/src/video/fbcon/SDL_fbvideo.c
--- SDL-1.2.5/src/video/fbcon/SDL_fbvideo.c	Mon Sep  2 16:42:15 2002
+++ SDL-1.2.5-freevo/src/video/fbcon/SDL_fbvideo.c	Fri Jan 10 00:27:10 2003
@@ -529,10 +529,18 @@
 	}
 
 	/* Enable mouse and keyboard support */
-	if ( FB_OpenKeyboard(this) < 0 ) {
-		FB_VideoQuit(this);
-		return(-1);
+	{
+		const char *sdl_nokbd;
+
+		sdl_nokbd = getenv("SDL_NOKEYBOARD");
+		if (!sdl_nokbd) {
+			if ( FB_OpenKeyboard(this) < 0 ) {
+				FB_VideoQuit(this);
+				return(-1);
+			}
+		}
 	}
+        
 	if ( FB_OpenMouse(this) < 0 ) {
 		const char *sdl_nomouse;
 
