*** mga_vid.c	Sat Aug  3 13:49:59 2002
--- mga_vid.c.new	Sat Aug  3 13:49:38 2002
***************
*** 1280,1286 ****
  			regs.beslumactl = (mga_brightness << 16) | ((mga_contrast+0x80)&0xFFFF);
  			mga_vid_write_regs(0);
  		break;
! 			
  	        default:
  			printk(KERN_ERR "mga_vid: Invalid ioctl\n");
  			return (-EINVAL);
--- 1280,1309 ----
  			regs.beslumactl = (mga_brightness << 16) | ((mga_contrast+0x80)&0xFFFF);
  			mga_vid_write_regs(0);
  		break;
! 		
! 		/* The following ioctl was added so that the _physical_ address of the
!                  * BES buffer can be obtained. This is needed when doing overlays from
!                  * a Video4Linux1 video capture card such as WinTV.
!                  * krister-freevo@kmlager.com
!                  */
!                 case MGA_VID_GET_BESADDR:
!                         if (!mga_src_base) {
!                                 printk(KERN_ERR "mga_vid: MGA_VID_GET_BESADDR "
!                                        "BES not setup!\n");
!                                 return(-EFAULT);
!                         }
!                         
!                         tmp = mga_mem_base + mga_src_base;
! 
!                         if (copy_to_user((uint32_t *) arg, &tmp, sizeof(uint32_t)))
!                         {
!                                 printk(KERN_ERR "mga_vid: MGA_VID_GET_BESADDR "
!                                        "failed copy %p to userspace %p\n",
!                                            &tmp, (uint32_t *) arg);
!                                 return(-EFAULT);
!                         }
!                 break;		
! 
  	        default:
  			printk(KERN_ERR "mga_vid: Invalid ioctl\n");
  			return (-EINVAL);
***************
*** 1645,1647 ****
--- 1668,1671 ----
  	unregister_chrdev(MGA_VID_MAJOR, "mga_vid");
  }
  
+ 
*** mga_vid.h	Sat Aug  3 13:49:59 2002
--- mga_vid.h.new	Sat Aug  3 13:49:38 2002
***************
*** 50,55 ****
--- 50,63 ----
  #define MGA_VID_GET_LUMA _IOR('J', 5, int)
  #define MGA_VID_SET_LUMA _IOR('J', 6, int)
  
+ /* The following ioctl was added so that the _physical_ address of the
+  *  BES buffer can be obtained. This is needed when doing overlays from
+  *  a Video4Linux1 video capture card such as WinTV.
+  *   krister-freevo@kmlager.com
+  */
+ #define MGA_VID_GET_BESADDR _IOR('J', 42, int)
+ 
+ 
  #define MGA_G200 0x1234
  #define MGA_G400 0x5678
  
