--- makelogos.py.orig	Tue Jul 16 11:42:40 2002
+++ makelogos.py	Tue Jul 16 11:42:53 2002
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/env python2
 #
 import xmltv
 import config
--- startup.py.orig	Tue Jul 16 11:43:13 2002
+++ startup.py	Tue Jul 16 11:43:56 2002
@@ -77,12 +77,12 @@
 # The remote control simulator, debug output in an Xterm
 remote_xterm = ('remote', 'xterm',
                     [ '-title', 'Freevo Remote Simulator', '-geom', '57x50',
-                      '-e', 'python', './rc_client/remote.py', '--remote=%(remote)s'])
+                      '-e', 'python2', './rc_client/remote.py', '--remote=%(remote)s'])
 
 # The remote control lirc interface, no debug output
 remote_quiet = ('remote', 'sh',
                      [ '-c',
-                       'python ./rc_client/remote.py --remote=lirc > /dev/null', ])
+                       'python2 ./rc_client/remote.py --remote=lirc > /dev/null', ])
 
 # The OSD framebuffer driver, no debug output
 osd_fb_quiet = ('osd_fb', 'sh', ['-c',
@@ -100,12 +100,12 @@
 # The Freevo main application, debug output in an Xterm
 freevo_main_xterm = ('freevo', 'xterm',
                      [  '-geom', '80x15', '-title', 'Freevo Main',
-                        '-e', 'python',
+                        '-e', 'python2',
                        './main.py', '--videotools=%(videotools)s'])
 
 # The Freevo main application, no debug output
 freevo_main_quiet = ('freevo', 'sh',
-                     [ '-c', 'python ./main.py '
+                     [ '-c', 'python2 ./main.py '
                        '--videotools=%(videotools)s > /dev/null'])
 
 #
--- strptime.py.orig	Tue Jul 16 11:54:30 2002
+++ strptime.py	Tue Jul 16 11:54:39 2002
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/env python2
 """Implements a pure Python version of strptime.
 
 
--- helpers/grabcover.py.orig	Tue Jul 16 11:45:14 2002
+++ helpers/grabcover.py	Tue Jul 16 11:45:27 2002
@@ -1,4 +1,4 @@
-#!/usr/bin/python
+#!/usr/bin/env python2
 #
 import os
 import urllib2
--- helpers/imdb.py.orig	Tue Jul 16 11:45:38 2002
+++ helpers/imdb.py	Tue Jul 16 11:45:51 2002
@@ -1,4 +1,4 @@
-#!/usr/bin/env python
+#!/usr/bin/env python2
 
 # usage:   imdb.py imdb-number out-filename moviefiles
 # example: imdb.py 0120915 swep1 star.wars.episode1.avi
--- freevo.orig	Tue Jul 16 11:51:54 2002
+++ freevo	Tue Jul 16 11:51:59 2002
@@ -5,4 +5,4 @@
 # $Id$
 #
 
-python startup.py $*
+python2 startup.py $*
