--- Imaging-1.1.3/Setup.in.orig	Mon Jul 15 15:21:16 2002
+++ Imaging-1.1.3/Setup.in	Mon Jul 15 15:21:33 2002
@@ -42,7 +42,7 @@
 
 _imagingtk  _imagingtk.c \
 	-IlibImaging Tk/tkImaging.c \
-	-I/usr/local/include -L/usr/local/lib -ltcl8.0 -ltk8.0 \
+	-I/usr/local/include -L/usr/local/lib -ltcl -ltk \
         -L/usr/X11R6/lib -lX11 
 
 # --------------------------------------------------------------------
