--- freevo	2003-11-10 15:05:19.000000000 -0400
+++ freevo	2003-11-17 20:39:30.000000000 -0400
@@ -407,25 +407,11 @@
 
 trap './freevo stop' 1 2 15
 
-# don't use -OO for Gentoo, it's hard to remove the pyo files
-# later (and it's a source for bugs
-if [ -e /etc/gentoo-release ]; then
-    OPTIMZE=""
-    if [ -e $FREEVO_PYTHON/main.pyo ]; then
-	echo "WARNING: you have Python pyo files in your system. They may"
-	echo "cause some bugs by using older version of some files."
-	echo "Please run 'find $FREEVO_PYTHON/.. -name \\*.pyo | xargs rm'"
-	echo
-    fi
-else
-    OPTIMZE="-OO"
-fi
-
 if [ "$1" = "start" ]; then
     shift
 
     # Start the main freevo application
-    $RUNAPP $PYTHON $OPTIMZE $FREEVO_PYTHON/main.py $@ > /dev/null 2>&1 &
+    $RUNAPP $PYTHON $FREEVO_PYTHON/main.py $@ > /dev/null 2>&1 &
     
     echo $! > $PID_FILE
     if [ `uname -s` != "FreeBSD" ]; then
