--- setup.py	2003-11-10 15:05:20.000000000 -0400
+++ setup.py	2003-11-10 20:13:37.000000000 -0400
@@ -28,19 +28,10 @@
     print 'of Freevo. Please run ./autogen.sh first'
     sys.exit(0)
 
-# add some files to Docs
-for f in ('BUGS', 'COPYING', 'ChangeLog', 'INSTALL', 'README'):
-    data_files.append(('share/doc/freevo-%s' % version.__version__, ['%s' % f ]))
-data_files.append(('share/doc/freevo-%s' % version.__version__, ['Docs/CREDITS' ]))
-
 # copy freevo_config.py to share/freevo. It's the best place to put it
 # for now, but the location should be changed
 data_files.append(('share/freevo', [ 'freevo_config.py' ]))
 
-# add docbook style howtos
-os.path.walk('./Docs/installation', docbook_finder, data_files)
-os.path.walk('./Docs/plugin_writing', docbook_finder, data_files)
-
 # i18n support
 i18n('freevo')
 
