diff -Naur libmpdemux/stream.h libmpdemux/stream.h.prot
--- mplayer-orig/libmpdemux/stream.h	Sat Jan  4 14:11:06 2003
+++ mplayer/libmpdemux/stream.h	Thu Feb 13 22:43:55 2003
@@ -206,6 +206,13 @@
 
 #ifdef USE_DVDREAD
 
+/*
+ * Override DVD protection (=DeCSS). Default == 1 (override = use DeCSS).
+ * This variable is set from the command line, and used in
+ * mplayer/libmpdvdkit2/libdvdcss.c.
+ */
+extern int dvdprotection_override;
+
 #ifdef USE_MPDVDKIT
 #if (USE_MPDVDKIT == 2)
 #include "../libmpdvdkit2/dvd_reader.h"
--- cfg-common.h	Thu Jun  5 22:30:32 2003
+++ cfg-common.h.prot	Thu Jun  5 22:27:52 2003
@@ -25,6 +25,9 @@
 	{"dvd", "MPlayer was compiled WITHOUT libdvdread support!\n", CONF_TYPE_PRINT, CONF_NOCFG, 0, 0, NULL},
 #endif
 	{"slang", &dvdsub_lang, CONF_TYPE_STRING, 0, 0, 0, NULL},
+        {"dvdprotection-override", &dvdprotection_override, CONF_TYPE_FLAG, 0, 0, 1, NULL},
+        {"nodvdprotection-override", &dvdprotection_override, CONF_TYPE_FLAG, 0, 1, 0, NULL},
+
 
 #ifdef HAVE_LIBCSS
         {"dvdauth", &dvd_auth_device, CONF_TYPE_STRING, 0, 0, 0, NULL},
--- libmpdvdkit2/libdvdcss.c	Thu Jun  5 22:33:10 2003
+++ libmpdvdkit2/libdvdcss.c.prot	Thu Jun  5 22:29:08 2003
@@ -105,6 +105,8 @@
 #include <unistd.h>
 #include <limits.h>
 
+#include "../mp_msg.h"
+
 #include "dvdcss.h"
 
 #include "common.h"
@@ -135,6 +137,10 @@
  */
 char * dvdcss_interface_2 = VERSION;
 
+/* Override DVD protection (=DeCSS). Default is on. */
+int dvdprotection_override = 1;
+
+
 /**
  * \brief Open a DVD device or directory and return a dvdcss instance.
  *
@@ -278,6 +284,20 @@
         }
     }
 
+    /* Check the global DVD protection override flag */
+    if ( dvdcss->b_scrambled && !dvdprotection_override )
+    {
+        static char err_msg[] = "The DVD is protected! "
+                                "Use -dvdprotection-override to override "
+                                "the protection, but only if legal "
+                                "in your jurisdiction!\n";
+
+        
+        /* The DVD is protected, and protection is not overridden by the user  */
+        mp_msg( MSGT_DVD, MSGL_ERR, err_msg);
+        return NULL;
+    }
+      
     /* If disc is CSS protected and the ioctls work, authenticate the drive */
     if( dvdcss->b_scrambled && dvdcss->b_ioctls )
     {
