Index: libmpdemux/cdda.c
===================================================================
RCS file: /cvsroot/mplayer/main/libmpdemux/cdda.c,v
retrieving revision 1.8
diff -u -r1.8 cdda.c
--- libmpdemux/cdda.c	15 Jan 2003 17:50:21 -0000	1.8
+++ libmpdemux/cdda.c	19 Feb 2003 05:55:21 -0000
@@ -182,8 +182,10 @@
   s->pos = p->sector*CD_FRAMESIZE_RAW;
   memcpy(s->buffer,buf,CD_FRAMESIZE_RAW);
 
-  if(p->sector == p->end_sector)
+  if((p->sector < p->start_sector) || (p->sector >= p->end_sector)) {
     s->eof = 1;
+    return;
+  }
 
   for(i=0;i<p->cd->tracks;i++){
 	  if(p->cd->disc_toc[i].dwStartSector==p->sector-1) {
@@ -206,6 +208,12 @@
   int sec;
   int current_track=0, seeked_track=0;
   int i;
+
+  
+  if(s->pos < 0) {
+    s->eof = 1;
+    return;
+  }
 
   sec = s->pos/CD_FRAMESIZE_RAW;
 //printf("pos: %d, sec: %d ## %d\n", (int)s->pos, (int)sec, CD_FRAMESIZE_RAW);
