--- mencoder.c	Mon Mar 10 12:12:11 2003
+++ mencoder.c.fb	Sat Apr 19 16:00:52 2003
@@ -363,7 +363,16 @@
   mp_msg_init();
   mp_msg_set_level(MSGL_STATUS);
   mp_msg(MSGT_CPLAYER,MSGL_INFO,"%s",banner_text);
+  
+  {
+    static char msg_freevo[] =
+      "\n\n\n\nFREEVO (http://www.freevo.org) VERSION!!!!!\n\n"
+      "DO NOT REQUEST SUPPORT FROM THE MPLAYER PROJECT WHEN "
+      "USING THIS VERSION OF MENCODER!!!!\n\n\n";
 
+    mp_msg(MSGT_CPLAYER,MSGL_INFO,msg_freevo);
+  }
+    
   /* Test for cpu capabilities (and corresponding OS support) for optimizing */
   GetCpuCaps(&gCpuCaps);
 #ifdef ARCH_X86
--- mplayer.c	Mon Mar 10 10:02:18 2003
+++ mplayer.c.fb	Sat Apr 19 16:01:11 2003
@@ -670,6 +670,16 @@
   mp_msg_set_level(MSGL_STATUS);
 
   mp_msg(MSGT_CPLAYER,MSGL_INFO,banner_text);
+  
+  {
+    static char msg_freevo[] =
+      "\n\n\n\nFREEVO (http://www.freevo.org) VERSION!!!!!\n\n"
+      "DO NOT REQUEST SUPPORT FROM THE MPLAYER PROJECT WHEN "
+      "USING THIS VERSION OF MPLAYER!!!!\n\n\n";
+
+    mp_msg(MSGT_CPLAYER,MSGL_INFO,msg_freevo);
+  }
+    
   /* Test for cpu capabilities (and corresponding OS support) for optimizing */
   GetCpuCaps(&gCpuCaps);
 #ifdef ARCH_X86
--- version.sh	Sun Apr  6 18:55:58 2003
+++ version.sh.fb	Sat Apr 19 16:04:12 2003
@@ -1,3 +1,3 @@
 #!/bin/sh
 
-echo "#define VERSION \"0.90rc5-$1 \"" >version.h
+echo "#define VERSION \"0.90-FREEVO \"" >version.h
