Index: vf.c
===================================================================
RCS file: /cvsroot/mplayer/main/libmpcodecs/vf.c,v
retrieving revision 1.51
diff -u -3 -p -r1.51 vf.c
--- vf.c	10 Oct 2002 01:09:23 -0000	1.51
+++ vf.c	16 Oct 2002 22:49:36 -0000
@@ -17,6 +17,7 @@
 
 extern vf_info_t vf_info_vo;
 extern vf_info_t vf_info_rectangle;
+extern vf_info_t vf_info_bmovl;
 extern vf_info_t vf_info_crop;
 extern vf_info_t vf_info_expand;
 extern vf_info_t vf_info_pp;
@@ -48,6 +49,7 @@ char** vo_plugin_args=(char**) NULL;
 // list of available filters:
 static vf_info_t* filter_list[]={
     &vf_info_rectangle,
+    &vf_info_bmovl,
     &vf_info_crop,
     &vf_info_expand,
     &vf_info_pp,
