Index: devices/oss.c
===================================================================
RCS file: /cvsroot/zapping/rte/mp1e/devices/oss.c,v
retrieving revision 1.6
diff -u -w -b -B -d -r1.6 oss.c
--- devices/oss.c	14 Dec 2002 00:43:44 -0000	1.6
+++ devices/oss.c	15 Feb 2003 05:58:11 -0000
@@ -329,17 +329,17 @@
 
 	ASSERT("get PCM rec source",
 	       IOCTL(fd, SOUND_MIXER_READ_RECSRC, &old_recsrc) == 0);
-	ASSERT("get PCM rec volume",
-	       IOCTL(fd, MIXER_READ(mix_line), &old_recvol) == 0);
+	/*ASSERT("get PCM rec volume",
+	       IOCTL(fd, MIXER_READ(mix_line), &old_recvol) == 0);S */
 
 	atexit(mix_restore);
 
 	ASSERT("set PCM rec source %d:%s",
 	       IOCTL(fd, SOUND_MIXER_WRITE_RECSRC,
 		     &recsrc) == 0, mix_line, sources[mix_line]);
-	ASSERT("set PCM rec volume %d%%",
+/*	ASSERT("set PCM rec volume %d%%",
 	       IOCTL(fd, MIXER_WRITE(mix_line),
-		     &recvol) == 0, mix_volume);
+		     &recvol) == 0, mix_volume); */
 
 	close(fd);
 
