diff -aur SDL-1.2.4/src/video/fbcon/SDL_fbevents.c SDL-1.2.4-freevo/src/video/fbcon/SDL_fbevents.c
--- SDL-1.2.4/src/video/fbcon/SDL_fbevents.c	Wed Mar  6 05:23:03 2002
+++ SDL-1.2.4-freevo/src/video/fbcon/SDL_fbevents.c	Mon Aug  5 23:15:52 2002
@@ -156,10 +156,40 @@
 	return((keyboard_fd >= 0) && (saved_kbd_mode >= 0));
 }
 
+static void
+tty_disable (void)
+{
+  int tty;
+
+
+  tty = open ("/dev/tty0", O_RDWR);
+  if(tty < 0) {
+    perror("Error can't open /dev/tty0");
+    exit (1);
+  }
+
+  if(ioctl (tty, KDSETMODE, KD_GRAPHICS) == -1) {
+    perror("Error setting graphics mode for tty");
+    close(tty);
+    exit (1);
+  }
+  
+  close(tty);
+
+}
+
+
 int FB_EnterGraphicsMode(_THIS)
 {
 	struct termios keyboard_termios;
+        int fd;
 
+        
+        if (getenv("SDL_NOKEYBOARD") != NULL) {
+          tty_disable ();
+          return(1);              /* XXX Do not touch the keyboard */
+        }
+        
 	/* Set medium-raw keyboard mode */
 	if ( (keyboard_fd >= 0) && !FB_InGraphicsMode(this) ) {
 
@@ -841,6 +871,10 @@
 	Uint32 screen_arealen;
 	Uint8 *screen_contents;
 
+        if (getenv("SDL_NOKEYBOARD") != NULL) {
+          return;        /* XXX Do not touch the keyboard */
+        }
+        
 	/* Figure out whether or not we're switching to a new console */
 	if ( (ioctl(keyboard_fd, VT_GETSTATE, &vtstate) < 0) ||
 	     (which == vtstate.v_active) ) {
@@ -891,6 +925,10 @@
 	int scancode;
 	SDL_keysym keysym;
 
+        if (getenv("SDL_NOKEYBOARD") != NULL) {
+          return;        /* XXX Do not touch the keyboard */
+        }
+        
 	nread = read(keyboard_fd, keybuf, BUFSIZ);
 	for ( i=0; i<nread; ++i ) {
 		scancode = keybuf[i] & 0x7F;
@@ -934,6 +972,10 @@
 	int max_fd;
 	static struct timeval zero;
 
+        if (getenv("SDL_NOKEYBOARD") != NULL) {
+          return;        /* XXX Do not touch the keyboard */
+        }
+        
 	do {
 		posted = 0;
 
diff -aur SDL-1.2.4/src/video/fbcon/SDL_fbvideo.c SDL-1.2.4-freevo/src/video/fbcon/SDL_fbvideo.c
--- SDL-1.2.4/src/video/fbcon/SDL_fbvideo.c	Wed Apr  3 15:45:25 2002
+++ SDL-1.2.4-freevo/src/video/fbcon/SDL_fbvideo.c	Mon Aug  5 23:17:53 2002
@@ -525,10 +525,18 @@
 	}
 
 	/* Enable mouse and keyboard support */
-	if ( FB_OpenKeyboard(this) < 0 ) {
-		FB_VideoQuit(this);
-		return(-1);
-	}
+        {
+          const char *sdl_nokbd;
+
+          sdl_nokbd = getenv("SDL_NOKEYBOARD");
+          if (!sdl_nokbd) {
+            if ( FB_OpenKeyboard(this) < 0 ) {
+              FB_VideoQuit(this);
+              return(-1);
+            }
+          }
+        }
+        
 	if ( FB_OpenMouse(this) < 0 ) {
 		const char *sdl_nomouse;
 
