diff -u -w -b -B -d -r syncfb.orig/Makefile syncfb/Makefile
--- syncfb.orig/Makefile	2001-01-23 12:26:59.000000000 -0500
+++ syncfb/Makefile	2003-07-25 12:11:18.000000000 -0400
@@ -1,8 +1,8 @@
+KERNELDIR = /usr/src/linux
+
 CFLAGS = -g -O2  -Wall -D__KERNEL__  -DMODULE -DSYNCFB_GENERIC_SUPPORT -DSYNCFB_MATROX_SUPPORT \
-	 -include /usr/src/linux/include/linux/modversions.h
+	 -include $(KERNELDIR)/include/linux/modversions.h -I$(KERNELDIR)/include
 
-%CFLAGS = -g -O2  -Wall -D__KERNEL__ -D__SMP__ -DMODULE \
-%	 -include /usr/src/linux/include/linux/modversions.h
 OBJ = syncfb_base.o syncfb_matrox.o syncfb_generic.o
 all: syncfb.o syncfb_test syncfb_off syncfbtv
 
diff -u -w -b -B -d -r syncfb.orig/syncfb.h syncfb/syncfb.h
--- syncfb.orig/syncfb.h	2001-03-04 19:23:16.000000000 -0500
+++ syncfb/syncfb.h	2003-07-25 12:09:40.000000000 -0400
@@ -10,7 +10,7 @@
 #include <linux/mm.h>
 #include <linux/string.h>
 #include <linux/errno.h>
-#include <linux/malloc.h>
+#include <linux/slab.h>
 #include <linux/pci.h>
 #include <linux/init.h>
 #include <linux/videodev.h>
diff -u -w -b -B -d -r syncfb.orig/syncfb_base.c syncfb/syncfb_base.c
--- syncfb.orig/syncfb_base.c	2001-03-04 19:23:16.000000000 -0500
+++ syncfb/syncfb_base.c	2003-07-25 12:08:55.000000000 -0400
@@ -27,8 +27,9 @@
 #undef VERSION
 #define VERSION(major,minor,patch) (((((major)<<8)+(minor))<<8)+(patch))
 
+MODULE_DESCRIPTION("SyncFB Matrox Video Driver");
 MODULE_AUTHOR("Matthias Oelmann <mao@well.com>");
-
+MODULE_LICENSE("GPL");
 
 /*
 	PROTOTYPES
